### PR TITLE
[FEAT][I] Customizable user highlight colors

### PR DESCRIPTION
**Pull request by davidschreiber**
Hi there, first of all: Thank you to everybody who contributed building Saros. This project seems like a promising platform for remote pair programming! This PR maybe comes out of the blue, but after trying out Saros (and so far really liking it), I quickly went ahead and fixed the hard coded coloring of user highlights (selection & contributions) which were made for light IDE themes (dark theme user here).

## What's inside this PR

* This PR replaces the previously hard-coded highlight colors inside ColorManager with user-customizable TextAttributes. It adds IntelliJ TextColorAttributes for every of the customizable colors which are 'selection' and 'contribution' colors for colors with IDs 0-4 and a separate set of default highlight colors (i.e. if no color could be matched using its ID).

* This PR also adds a color preference screen to the IDE, so that the color attributes can be modified inside  Preferences -> Editor -> Color Scheme -> Saros. 

* This PR adds two default color schemes configurations for the IntelliJ "Default" and "Darcula" schemes.

Here's a screenshot of the new IDE color scheme preferences:

Preferences

## Noteworthy

*UPDATE: Nevermind, found this PR https://github.com/saros-project/saros/pull/395*. ~After finding out more about the Java code formatting rules of the Saros project, I've installed the `google-java-format` version `1.7.0.1` inside my IntelliJ. After reformatting files, I still got Travis CI formatting errors. I noticed that Travis uses version `1.6.0` of the formatter, which seems to slightly have altered formatting rules (a single line was changed after formatting with 1.6.0). Are there any plans on upgrading to the latest formatter?~

### CODE REVIEW

**Comment by tobous** [comment]
Hey David, thanks for your initiative. I only skimmed the code so far but it look promising. I will have a more detailed look at it in the next couple of days.

File: intellij/src/saros/intellij/editor/colorstorage/ColorManager.java [hunk 1]
```diff
@@ -1,44 +1,157 @@
...
+    for (int i = 0; i < USER_COUNT; i++) {
+      builder.add(new SarosUserColorKeys(i));
+    }
+    USER_COLOR_KEYS = builder.build();
+  }
 
   private ColorManager() {}
 
   /**
    * Returns the color for the given userID. Returns the DEFAULT_COLOR when there is no color for
```
[the remark this answers is not in the document]
Comment (davidschreiber) [reply intellij/src/saros/intellij/editor/colorstorage/ColorManager.java]: Got it! That makes sense. I'll update the PR to reflect this structure.

File: intellij/src/saros/intellij/editor/annotations/AnnotationManager.java [hunk 2]
```diff
@@ -687,24 +688,24 @@ private RangeHighlighter addRangeHighlighter(
       return null;
     }
 
-    Color color;
+    // Retrieve color keys based on the color ID selected by this user. This will automatically
+    // fall back to default colors, if no colors for the given ID are available.
+    final ColorKeys colorKeys = ColorManager.getColorKeys(user.getColorID());
```
Comment (davidschreiber) [inline intellij/src/saros/intellij/editor/annotations/AnnotationManager.java]: I noticed the discrepancy of the `User#getColorID()` method name casing and the `ColorKeys#getId()` method. According to the Google Java stylebook, `getId()` would be the correct name: https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

Is this divergence from the style intentional, or is it an outdated format? Should I fix the name of the `ColorKeys#getId()` method, or should we keep it and create an issue for fixing the remaining code?

**Comment by fzieris** [comment]
First of all: Welcome, David. Thank you very much for getting in touch!

I'd like to provide some background information on the user colors in Saros.
(I assume that Saros/IntelliJ uses the same colors as Saros/Eclipse.)
In 2011, we settled on these particular five hard-coded colors based on how well users with different forms of color blindness would be able to read Java source code in Eclipse's standard syntax highlighting and to tell different users apart (obviously less of a concern for two-user sessions, where each user only sees the color of her partner).

The [Master's thesis](http://www.inf.fu-berlin.de/inst/ag-se/theses/Kahlert11-saros-out-of-box-experience.pdf) which dealt with this issue (among with many others) is written in German; the user colors are discussed on pages 181-192.
For an optical intuition of how up to 10% of humans would see this color scheme in the Eclipse IDE see pages 190-191: The first row represents "normal" vision, the other rows imitate various forms of color blindness.
For comparison, refer to pages 186-187 where a naive color scheme (in fact the old Saros color scheme) is subject to the same conditions. Reading the source code under the highlight may get difficult then.

With a different background color (e.g., dark grey) and syntax highlighting style, things get more complicated.
All I'm saying is that an informed design decision is not easy here and that users maybe should not be enabled to pick colors from the whole RGB color space.

**Comment by davidschreiber** [comment]
Hi Franz,

Thanks for providing insights on the original color finding process. For this PR, I have kept the original hard-coded colors and transferred them to the `SarosDefault.xml` scheme. This means that by default, the same colors will be used as on Eclipse. 

In addition, I added a set of "dark themed" colors. However, I haven't performed any accessibility research about the picked colors, or any other data driven color finding in that regard, but generated those colors by taking the original colors and adapting their contrast values for subjective better readability and experience in the Darcula theme.

If we aim for improving the quality of colors for users affected by color blindness, I would suggest that we add specific color sets for those users. IntelliJ allows plugins to specify alternative colors for deuteranopia and protanopia, and allows users to enable these colors inside their IDE's preferences.

Inside the color scheme file, these values would be provided like so:

```xml
<option name="SAROS_TEXT_SELECTION_1">
  <value>
    <option name="BACKGROUND" value="BFBB82" deuteranopia="cccccc" protanopia="cccccc"/>
  </value>
</option>
```

(Please note that the color values are simply meant as placeholders here).

**Comment by fzieris** [comment]
Very well, sounds reasonable. I just wanted to add my two cents and kindly ask @m273d15 and @tobous to reach a verdict, as I withdrew from Saros development last summer. I just wasn't sure if anyone else still remembered that there was some thorough thought behind the current color scheme.

**Comment by tobous** [comment]
I think keeping the current colors as the preset default is the right way to go. In the future, we could discuss to use the specific color-blind options and choose some "nicer" colors as defaults, but as "nicer colors" is *very* subjective, I don't think this has a very high priority for now.

As to the defined presets:
@davidschreiber For me, the drop down menu for color schemes on the Saros color page also include a "High contrast" option by default which seems to use the same colors as Darcula. Have you had a look at these colors as well? Maybe it would be useful to also add have an adjusted default config for this scheme.

Concerning the question whether we should restrict the color choices:
I don't think we should restrict the user in their choice of annotation color. With good default values that should be usable by almost everybody out of the box, I think the user should be allowed to shoot themselves in the foot if they really want to. Or in other words, if a user thinks using white annotations on a white background is a good idea, that is not really our problem. :smile:
A fully customizable color range is especially useful as of IntelliJ 2019.1 as they introduced [fully customizable IDE color schemes](https://blog.jetbrains.com/idea/2019/03/brighten-up-your-day-add-color-to-intellij-idea/), meaning it is nearly impossible for us to find some kind of restriction that works with every imaginable color scheme.

**Comment by tobous** [comment]
@davidschreiber Are you still interested in working on this PR?

**Comment by davidschreiber** [comment]
@tobous Sorry for letting this become stale. Yes, I still intend to resolve the outstanding issues and contribute this in a final state. I'm plan to address this in the upcoming days.

**Comment by tobous** [comment]
Superseded by #632. As this is an external PR, I did not have the permissions to modify the contents of the PR (as it is based on a fork of the project). To avoid this, I rebased and squashed the commits contained in the PR and included the result in the new PR.

@davidschreiber As I wanted to include this functionality in the next release, I just finished it myself. I hope this does not other you.

**Pull request closed without merging**